onnxruntime/core/providers/cpu/math/einsum_utils/einsum_auxiliary_ops.cc

```diff
@@ -251,6 +251,28 @@ std::unique_ptr<Tensor> Diagonal(const Tensor& input, int64_t dim_1, int64_t dim
 
 }  // namespace DeviceHelpers
 
+// This helps decide if we need to reshape instead of transposing.
+bool IsReshapeRequired(const std::vector<int64_t>& input_dims, const std::vector<size_t>& permutation) {
+  ORT_ENFORCE(input_dims.size() == permutation.size(), "The rank of the input must match permutation size for Transpose/Reshape.");
```

Comment:
Is anything different between the implementation of IsReshapeRequired here vs. IsReshape in tranpose.cc? Can we just have a single version of the logic checking the dims vs perms?

Reply:
Just to precise, this PR includes another one: #5550. This one is an addition to save one allocation in operator Einstein. I was hoping the first one to be reviewed and then discuss about the best way to save this allocation and to measure the gain obtained with this change. I'll propagate the changes into #5550.